bittide/src/Bittide/Transceiver/ResetManager.hs

```diff
@@ -0,0 +1,186 @@
+  -- will reset if the transmit side resets - see 'rxFullRetries'.
+  , rxFullRetries :: Unsigned 32
+  -- ^ How many times 'rxRetries' overflowed. I.e., how many times 'RxWait' moved
+  -- the state machine back to 'StartTx'.
+  , failAfterUps  :: Unsigned 32
+  -- ^ How many times the link failed when in the 'Monitor' state - i.e., after
+  -- detecting it fully worked. This usually happens if the other side drops its
+  -- link because it tried resetting its receive side too many times - see
+  -- 'rxFullRetries'.
+  }
+  deriving (Generic, NFDataX)
+
+-- | Configuration for 'resetManager'
+--
+-- Develop notes: the current API is a balance between the tightest possible
+-- hardware and the most flexible API. We could make the API more flexible by
+-- exposing 'MaxTxTimeoutMs' and friends as type parameters, but that severely
+-- impacts verbosity and readability.
+data Config = Config
+  { txTimeoutMs :: MaxTxTimeoutMs
+  -- ^ Number of milliseconds to wait for the transmit side to be ready, before
+  -- resetting the transmit subsystem.
+  , rxTimeoutMs :: MaxRxTimeoutMs
+  -- ^ Number of milliseconds to wait for the receive side to be ready, before
+  -- resetting the receiver subsystem.
+  , rxRetries :: MaxRxRetries
+  -- ^ Number of times to retry the receive side before resetting the transmit
+  -- side as well.
+  }
+  deriving (Generic, NFDataX)
+
+-- | Default configuration for 'resetManager'
+--
+-- XXX: Current timeout values for 'TxWait' and 'RxWait' are chosen arbitrarily.
+--      We should investigate what these values should be for quick bring-up.
+defConfig :: Config
+defConfig = Config
+  { txTimeoutMs = 1
+  , rxTimeoutMs = 5
+  , rxRetries = 8
+  }
+
+-- | Bringing up the transceivers is a stochastic process - at least, that is
+-- what Xilinx reference designs make us believe. We therefore retry a number of
+-- times if we don't see sensible data coming in. See the individual constructors
+-- and 'resetManager' for more information.
+data State dom
+  = StartTx Statistics
+  -- ^ Reset everything - transmit and receive side
+  | StartRx Statistics
+  -- ^ Reset just the receive side
+  | TxWait Statistics (MaxTxTimeoutMs, IndexMs dom 1)
+  -- ^ Wait for the transmit side to report it is done. After /n/ milliseconds
+  -- (see type) it times out, moving to 'StartTx'.
+  | RxWait Statistics (MaxRxTimeoutMs, IndexMs dom 1)
+  -- ^ Wait for the receive side to report it is done _and_ that it can predict
+  -- the data coming from the other side. After /n/ milliseconds (see type) it
+  -- times out. Depending on the value of 'ResetStat's 'rxRetries' it will
+  -- either reset both the receive and the transmit side, or just the receive
+  -- side. If all is well though, move on to 'Monitor'.
+  | Monitor Statistics
+  -- ^ Wait till the end of the universe, or until a link goes down - whichever
+  -- comes first. In case of the latter, the state machine moves to 'StartTx'.
+  deriving (Generic, NFDataX)
```

Comment:
If `Statistics` is part of every state anyway, then it is more convenient to leave it from the type an run the state machine on `(Statistics, State)` instead.